plugins/src/client/plugin.rs

```diff
@@ -38,33 +38,35 @@ where
     // to get one from Self::InputWrapper, potentially falliably
     Self: Callable + TryFrom<Self::InputWrapper>,
     // InputWrapper must be deserializable and describable
-    Self::InputWrapper: JsonSchema + Sized + for<'a> Deserialize<'a>,
+    Self::InputWrapper: Sized + for<'a> Deserialize<'a> + HasSapioModuleSchema,
     // read as: The return type of CallableType::try_from(self) is
     // Result<CallableType, X>, where X must be able to x.into() a
     // CompilationError.
     CompilationError: From<<Self as TryFrom<Self::InputWrapper>>::Error>,
     // We must be able to serialize/describe the outputs
-    Self::Output: Serialize + JsonSchema,
+    Self::Output: Serialize + HasSapioModuleSchema,
 {
     /// A type which wraps Self, but can be converted into Self.
     type InputWrapper;
     /// gets the jsonschema for the plugin type, which is the API for calling create.
     fn get_api_inner() -> *mut c_char {
-        encode_json(&API::<CreateArgs<Self::InputWrapper>, Self::Output>::new())
+        encode_boundary_repr(&API::<CreateArgs<Self::InputWrapper>, Self::Output>::get_schema())
     }
 
     /// creates an instance of the plugin from a json pointer and outputs a result pointer
     unsafe fn create(p: *mut c_char, c: *mut c_char) -> *mut c_char {
         let res = Self::create_result(p, c).map_err(|e| e.to_string());
-        encode_json(&res)
+        encode_boundary_repr(&res)
     }
 
     /// creates an instance of the plugin from a json pointer and outputs a typed result
     unsafe fn create_result(
         p: *mut c_char,
         c: *mut c_char,
     ) -> Result<Self::Output, CompilationError> {
+        todo!("figure out how to deal with c strings");
         let s = CString::from_raw(c);
+        todo!("figure out how to deal with c strings");
```

Comment:
i think in general in this commit it's not clear _why_ so much of the serialization logic is breaking so it has to be todo!'d out.